### PR TITLE
Per-category loss summary

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
This PR adds support to do per-category loss summary, as shown in the examples below:

image
image

### CODE REVIEW

File: alf/utils/summary_utils.py [hunk 1]
```diff
@@ -165,6 +165,39 @@ def add_nested_summaries(prefix, data):
             alf.summary.scalar(name, elem)
 
 
+@_summary_wrapper
+def summarize_per_category_loss(loss_info: LossInfo):
```
Comment (emailweixu) [inline alf/utils/summary_utils.py]: perhaps can add an optional argument: ``label_name: Optional[List[str]]=None`` to use label name for the summary name. ``label_name`` can be configured using ``alf.config()``

Comment (Haichao-Zhang) [reply alf/utils/summary_utils.py]: Added label_label arguments and now the summarization could be more informative, as shown in the updated PR description.

File: alf/utils/summary_utils.py [hunk 2]
```diff
@@ -165,6 +166,59 @@ def add_nested_summaries(prefix, data):
...
+                                summarize_count=False,
+                                label_name: Optional[List[str]] = None):
+    """Add summary about each category of the unaggregated ``loss_info.loss``
+    of the shape (T, B), or (B, ) by partationing it according to
+    ``loss_info.batch_label``, which has the same shape as ``loss_info.loss``.
+    It also creates summarization of the number of samples encountered
+    for each category.
+
+    Args:
+        loss_info (LossInfo): do per-category summarization if
```
Comment (emailweixu) [inline alf/utils/summary_utils.py]: No need to annotate type if type hints are provided in the function signature.

File: alf/utils/summary_utils.py [hunk 3]
```diff
@@ -165,6 +166,59 @@ def add_nested_summaries(prefix, data):
             alf.summary.scalar(name, elem)
 
 
+@_summary_wrapper
+@alf.configurable
+def summarize_per_category_loss(loss_info: LossInfo,
+                                summarize_count=False,
+                                label_name: Optional[List[str]] = None):
```
Comment (emailweixu) [inline alf/utils/summary_utils.py]: label_name -> label_names

File: alf/utils/summary_utils.py [hunk 4]
```diff
@@ -165,6 +166,59 @@ def add_nested_summaries(prefix, data):
...
+    """Add summary about each category of the unaggregated ``loss_info.loss``
+    of the shape (T, B), or (B, ) by partationing it according to
```
Comment (emailweixu) [inline alf/utils/summary_utils.py]: partationing -> partitioning

File: alf/utils/summary_utils.py [hunk 5]
```diff
@@ -165,6 +166,59 @@ def add_nested_summaries(prefix, data):
...
+        # (T, B) -> (T * B, )
+        loss = loss_info.loss.reshape(-1)
+        batch_label = loss_info.batch_label.int().reshape(-1)
+        labels = torch.unique(batch_label)
+        labels = labels.tolist()
+
+        for label in labels:
+            subset_indices = (
+                batch_label == label).nonzero().reshape(-1).long()
+            subset_loss = torch.index_select(loss, 0, subset_indices)
```
Comment (emailweixu) [inline alf/utils/summary_utils.py]: A more readable form: `loss[bach_label==label]`

**Review by emailweixu — DISMISSED** [review]

File: alf/utils/summary_utils.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/utils/summary_utils.py]: Changed

File: alf/utils/summary_utils.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/utils/summary_utils.py]: Removed unnecessary type annotations.

File: alf/utils/summary_utils.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/utils/summary_utils.py]: typo fixed

File: alf/utils/summary_utils.py (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/utils/summary_utils.py]: Simplified as suggested.

**Review by emailweixu — APPROVED** [review]
